source/common/buffer/watermark_buffer.cc

```diff
@@ -62,6 +62,29 @@ SliceDataPtr WatermarkBuffer::extractMutableFrontSlice() {
   return result;
 }
 
+// Adjust the reservation size based on space available before hitting
+// the high watermark to avoid overshooting by a lot and thus violating the limits
+// the watermark is imposing.
+Reservation WatermarkBuffer::reserve(uint64_t preferred_length) {
+  uint64_t adjusted_length = preferred_length;
+
+  if (high_watermark_ > 0 && preferred_length > 0) {
+    const uint64_t current_length = OwnedImpl::length();
+    if (current_length >= high_watermark_) {
+      // Always allow a read of at least some data. The API doesn't allow returning
+      // a zero-length reservation.
+      adjusted_length = OwnedSlice::default_slice_size_;
+    } else {
+      const uint64_t available_length = high_watermark_ - current_length;
```

Comment:
https://github.com/envoyproxy/envoy/pull/14333 could help.  We could use some benchmarks.

Reply:
I'm trying to wrap my head around the interactions of the watermarkbuffer, connection, and transport socket on this. My understanding is that the transport socket will keep reading until either there's no more data, or until `shouldDrainReadBuffer()` which is `read_buffer_limit_ > 0 && read_buffer_->length() >= read_buffer_limit_`. 

The original code did this in chunks of always 16k. I think the new code behaves roughly the same in terms of how many bytes will be read, except it will do up to 128k chunks, but only up to 16k more than the limit (same result as old code).

Except the difference might be on the boundary case, where the available length is evenly divisible by 16k. 

Does the code as written here have a net behavior change regarding triggering the watermarks?

Reply:
I think #14333 is needed to get this change to work as expected.  Otherwise this change always ends up reading an extra 16kb beyond what is configured and the connection goes down the known-to-be-slower readDisable(true)/readDisable(false) path.  The issue is that there's no precise low-level tests that validate that we end up avoiding the readDisable path.

Reply:
I'll have to see if I can make some time to help do the cleanup in #14333 and unblock this.